services/src/test/scala/cromwell/pipeline/service/ProjectFileServiceTestImpl.scala

```diff
@@ -0,0 +1,57 @@
+package cromwell.pipeline.service
+import cromwell.pipeline.datastorage.dto._
+import cromwell.pipeline.model.wrapper.UserId
+
+import java.nio.file.Path
+import scala.concurrent.Future
+
+class ProjectFileServiceTestImpl(projectFiles: Seq[ProjectFile], testMode: TestMode) extends ProjectFileService {
+
+  override def validateFile(fileContent: ProjectFileContent): Future[Either[ValidationError, Unit]] =
+    testMode match {
+      case WithException(exc: ValidationError) => Future.successful(Left(exc))
+      case WithException(exc)                  => Future.failed(exc)
+      case _                                   => Future.successful(Right(Unit))
+    }
+
+  override def uploadFile(
+    projectId: ProjectId,
+    projectFile: ProjectFile,
+    version: Option[PipelineVersion],
+    userId: UserId
+  ): Future[Either[VersioningException, Unit]] =
+    testMode match {
+      case WithException(exc: VersioningException) => Future.successful(Left(exc))
+      case WithException(exc)                      => Future.failed(exc)
+      case _                                       => Future.successful(Right(Unit))
+    }
+
+  override def getFile(
+    projectId: ProjectId,
+    path: Path,
+    version: Option[PipelineVersion],
+    userId: UserId
+  ): Future[ProjectFile] =
+    testMode match {
+      case WithException(exc) => Future.failed(exc)
+      case _                  => Future.successful(projectFiles.head)
```

Comment:
Maybe we want to take the file at least by name? 
Then we can write something like:
```suggestion
      case _                  => Future.successful(projectFiles.filter(_.path == path).take(1))
```

Reply:
I thought that we will create a new testImpl for every testcase. So to that testImpl we will pass objects which we want to return. And if we want to receive one specific file we will pass just that single file. We just don't need to pass a whole bunch to choose from. Please, correct me if I'm wrong.

Reply:
I just think this will be more expected behavior from a test implementation. It's just that if we decide to write a test where there will be several elements, then we cannot be sure that the one we need will be returned to us.

Reply:
Fixed, thanks

Reply:
Ok.